cpp/include/raft/core/mdspan.hpp

```diff
@@ -22,4 +15,212 @@
  */
 #pragma once
 
-#include <raft/thirdparty/mdspan/include/experimental/mdspan>
\ No newline at end of file
+#include <raft/core/error.hpp>
+#include <raft/core/mdspan_types.hpp>
+
+#include <raft/core/detail/accessor_mixin.hpp>
+#include <raft/core/detail/mdspan_util.cuh>
+
+#include <raft/thirdparty/mdspan/include/experimental/mdspan>
+
+namespace raft {
+
+template <typename ElementType,
+          typename Extents,
+          typename LayoutPolicy   = layout_c_contiguous,
+          typename AccessorPolicy = std::experimental::default_accessor<ElementType>>
+using mdspan = std::experimental::mdspan<ElementType, Extents, LayoutPolicy, AccessorPolicy>;
+
+/**
+ * Ensure all types listed in the parameter pack `Extents` are integral types.
+ * Usage:
+ *   put it as the last nameless template parameter of a function:
+ *     `typename = ensure_integral_extents<Extents...>`
+ */
+template <typename... Extents>
+using ensure_integral_extents = std::enable_if_t<std::conjunction_v<std::is_integral<Extents>...>>;
+
+/**
+ * @\brief Template checks and helpers to determine if type T is an std::mdspan
+ *         or a derived type
+ */
+
+template <typename ElementType, typename Extents, typename LayoutPolicy, typename AccessorPolicy>
+void __takes_an_mdspan_ptr(mdspan<ElementType, Extents, LayoutPolicy, AccessorPolicy>*);
+
+template <typename T, typename = void>
+struct is_mdspan : std::false_type {
+};
+template <typename T>
+struct is_mdspan<T, std::void_t<decltype(__takes_an_mdspan_ptr(std::declval<T*>()))>>
```

Comment:
Would you consider a simpler design that just takes a single template parameter T and specializes on it?

```c++
template<typename T>
struct is_mdspan : std::false_type {};

template<typename ElementType, typename Extents, typename Layout, typename Accessor>
struct is_mdspan<mdspan<ElementType, Extents, Layout, Accessor>> : std::true_type {};
```